README.md
# j_realty_api: Knowing the real estate transaction prices in Japan.

## What is it?

The j_realty_api is a wrapper for an API that allows you to retrieve real estate transaction prices in Japan, utilizing the [Real Estate Transaction Price Information Retrieval API](https://www.land.mlit.go.jp/webland/api.html) provided by Japan's Ministry of Land, Infrastructure, Transport and Tourism. It makes the use of the API more convenient. It supports both Japanese and English APIs.

A feature of the tool is that it includes a tool to obtain prefecture codes and municipality codes, making it easier to search for data.

j_realty_apiは、日本の国土交通省が出している[不動産取引価格情報取得API](https://www.land.mlit.go.jp/webland/api.html)を使い、日本の不動産取引価格を取得できる、APIのラッパーです。APIの利用が容易になります。日本語API、英語APIの両方に対応しています。

ツールの特徴は、都道府県コードと市町村コードを取得できるツールを加えて、データを探しやすくした点です。

## Installation

- インストール方法

```sh
pip install j_realty_api
```

```sh
pip install git+https://github.com/mazarimono/j_realty_api.git
```

## How to Use
### 使い方

- English
    - [sample notebook](https://github.com/mazarimono/j_realty_api/tree/main/samples/sample_j_realty_en.ipynb)

```python
# English
from j_realty_api import j_realty_en as jrn

j = jrn.CityCode('Kyoto')
pref_code = j.pref_code
city_code = j.city_code('kyoto')
pt = jrn.PropTransactions(pref_code, city_code, 20223, 20224)
df = pt.get_data()
# Can get Prices with forex(USD, EUR, CNY) (from ver0.2)
df_forex = pt.to_forex(df, 'USD')
```

- 日本語
- [sample notebook](https://github.com/mazarimono/j_realty_api/tree/main/samples/sample_j_realty_jp.ipynb)

```python
from j_realty_api import j_realty_jp as jrp

j = jrp.CityCode('京都')
pref_code = j.pref_code
city_code = j.city_code('京都')
pt = jrp.PropTransactions(pref_code, city_code, 20223, 20224)
df = pt.get_data()
```


## Licence
MIT

CHANGELOG.md
## 開発メモ


### ver0.2

- add to_forex method to j_realty_en
- 


### 20231020

- released version_0.1
- this package assists get japanese real estate transaction prices from api(https://www.land.mlit.go.jp/webland/api.html).
- バージョン0.1のリリース
- 国交省がリリースしている不動産取引価格情報取得APIのラッパー
    


tests/test.py
from j_realty_api import j_realty_en as je

cc = je.CityCode("kyoto")
pref_code = cc.pref_code
city_code = cc.city_code("kyoto")
pt = je.PropTransactions(
    pref_code,
    city_code,
)

df = pt.get_data()
df.to_csv("./tests/data/test.csv")
print(df.info())


tests/check.py
import json

with open('./j_realty_api/data', 'r') as f:
    data = json.load(f)

print(data)



tests/test1.py
from j_realty_api import j_realty_jp



prop_api/prop_api.py
import json
import math
from dataclasses import dataclass
from typing import Union

import numpy as np
import pandas as pd
import pkg_resources
import requests

CITYSEARCH_URL = "https://www.land.mlit.go.jp/webland/api/CitySearch"
TRADESEARCH_URL = "https://www.land.mlit.go.jp/webland/api/TradeListSearch"
PREF_CODE_JSON_PATH = pkg_resources.resource_filename(
    "prop_api", "data/japan-pref-code.json"
)


@dataclass
class CityCode:

    """
    都道府県名を渡すと、その都道府県の市区町村名とコードの格納された
    データが返される。
    市区町村名を指定すると、市区町村コードが返される
    """

    pref_name: str
    base_path: str = CITYSEARCH_URL

    @property
    def pref_code(self) -> str:
        """
        都道府県名からコードを探す
        """
        with open(PREF_CODE_JSON_PATH, "r", encoding="utf-8") as f:
            pref_code = json.load(f)
        for p_name, p_code in pref_code.items():
            if p_name.startswith(self.pref_name):
                return p_code
        raise Exception(f"No pref_code found for {self.pref_name}")

    @property
    def city_json(self):
        r = requests.get(self.base_path, params={"area": self.pref_code})
        return r.json()

    def city_code(self, city_name: str):
        for d in self.city_json.get("data", []):
            if d["name"].startswith(city_name):
                return d["id"]
        raise Exception(f"No city_code found for {city_name}.")


@dataclass
class PropTransactions:

    """
    area: str
        都道府県コード
    city: str
        市区町村コード
    from_dt: str
    to_dt: str
        'YYYYQ' の形で表現される。Q は四半期
        形式はYYYYN（数字5桁）
        YYYY … 西暦
        N … 1～4（1:1月～3月、2:4月～6月、3:7月～10月、4:11月～12月）
        ※20053（平成17年第３四半期）より指定可能
    """

    area: str
    city: str
    from_dt: int = 20221
    to_dt: int = 20223
    base_url: str = TRADESEARCH_URL

    def get_data(self):
        params = {
            "from": self.from_dt,
            "to": self.to_dt,
            "area": self.area,
            "city": self.city,
        }
        r = requests.get(self.base_url, params=params)
        if r.status_code == 200:
            df = pd.DataFrame(r.json()["data"])
            df = self._prepro_df(df)
            df["BuildingYear"] = df["BuildingYear"].map(self._wareki_to_seireki)
            df["Period"] = df["Period"].map(self._convert_to_quarter)
            return df
        else:
            raise Exception(f"Status_code: {r.status_code}")

    def _prepro_df(self, df: pd.DataFrame) -> pd.DataFrame:
        """
        データを使いやすいように前処理するメソッド
        """
        int_col_list = [
            "TradePrice",
            "Area",
            "CoverageRatio",
            "FloorAreaRatio",
            "PricePerUnit",
            "UnitPrice",
        ]
        float_col_list = ["Frontage", "Breadth"]

        def _to_int(x: str):
            """
            数値型に変更する関数
            nanをはじきたい
            """
            try:
                x = int(x)
                return x
            except:
                return x

        for col in int_col_list:
            if col == "Area":
                df[col] = df[col].map(lambda x: x.replace("㎡以上", ""))
            df[col] = df[col].map(_to_int)

        for col in float_col_list:
            df[col] = df[col].astype(float, errors="ignore")

        return df

    def _wareki_to_seireki(self, x: Union[str, float]) -> int:
        if not x or isinstance(x, float) and np.isnan(x):
            return np.nan
        try:
            if x[:2] == "令和":
                year = int(x[2:-1])
                wareki = 2018 + year
                return wareki
            elif x[:2] == "平成":
                year = int(x[2:-1])
                wareki = 1988 + year
                return wareki
            elif x[:2] == "昭和":
                year = int(x[2:-1])
                wareki = 1925 + year
                return wareki
        except ValueError:
            return np.nan

    def _convert_to_quarter(self, input_str: str) -> pd.Period:
        # 年と四半期の情報を抽出
        year = int(input_str.split("年")[0])
        quarter = int(input_str.split("第")[1][0])
        return pd.Period(f"{year}Q{quarter}")


if __name__ == "__main__":
    t = CityCode("京都")
    pref_code = t.pref_code
    city_code = t.city_code("京都市")
    kt = PropTransactions(pref_code, city_code, 20223, 20231)
    df = kt.get_data()
    print(df.head())


j_realty_api/j_realty_jp.py
import json
import math
from dataclasses import dataclass
from typing import Union

import numpy as np
import pandas as pd
import pkg_resources
import requests

# API from 国土交通省
# web: https://www.land.mlit.go.jp/webland/api.html

CITYSEARCH_URL = "https://www.land.mlit.go.jp/webland/api/CitySearch"
TRADESEARCH_URL = "https://www.land.mlit.go.jp/webland/api/TradeListSearch"
PREF_CODE_JSON_PATH = pkg_resources.resource_filename(
    "j_realty_api.j_realty_jp", "data/pref-code-jp.json"
)


@dataclass
class CityCode:

    """
    都道府県名を渡すと、その都道府県の市区町村名とコードの格納された
    データが返される。
    市区町村名を指定すると、市区町村コードが返される
    """

    pref_name: str
    base_path: str = CITYSEARCH_URL

    @property
    def pref_code(self) -> str:
        """
        都道府県名からコードを探す
        """
        with open(PREF_CODE_JSON_PATH, "r", encoding="utf-8") as f:
            pref_code_json = json.load(f)
        for p_name, p_code in pref_code_json.items():
            if p_name.startswith(self.pref_name):
                return p_code
        raise Exception(f"No pref_code found for {self.pref_name}")

    @property
    def city_json(self) -> dict[str]:
        """
        指定した都道府県の市町村コードのJSON
        """
        r = requests.get(self.base_path, params={"area": self.pref_code})
        return r.json()

    def city_code(self, city_name: str) -> str:
        """
        city_nameに対応したcitycodeが返される
        """
        for d in self.city_json.get("data", []):
            if d["name"].startswith(city_name):
                return d["id"]
        raise Exception(f"No city_code found for {city_name}.")


@dataclass
class PropTransactions:

    """
    APIから各パラメータに見合ったデータを取得する

    pref_code: str
        都道府県コード
    city_code: str
        市区町村コード
    from_dt: str
    to_dt: str
        'YYYYQ' の形で表現される。Q は四半期
        形式はYYYYN（数字5桁）
        YYYY … 西暦
        N … 1～4（1:1月～3月、2:4月～6月、3:7月～10月、4:11月～12月）
        ※20053（平成17年第3四半期）より指定可能
    """

    pref_code: str
    city_code: str
    from_dt: int = 20222
    to_dt: int = 20223
    base_url: str = TRADESEARCH_URL

    def get_data(self) -> pd.DataFrame:
        """
        指定された都道府県コード、市町村コード、期間の
        不動産取引価格情報を取得する。
        データはJSONをDataFrameにして返す。
        Returns:
            pd.DataFrame
        """
        params = {
            "from": self.from_dt,
            "to": self.to_dt,
            "area": self.pref_code,
            "city": self.city_code,
        }
        r = requests.get(self.base_url, params=params)
        if r.status_code == 200:
            df = pd.DataFrame(r.json()["data"])
            df = self._prepro_df(df)
            df["BuildingYear"] = df["BuildingYear"].map(self._wareki_to_seireki)
            df["Period"] = df["Period"].map(self._convert_to_quarter)
            return df
        else:
            raise Exception(f"Status_code: {r.status_code}")

    def _prepro_df(self, df: pd.DataFrame) -> pd.DataFrame:
        """
        データを使いやすいように前処理するメソッド
        """
        int_col_list = [
            "TradePrice",
            "Area",
            "CoverageRatio",
            "FloorAreaRatio",
            "PricePerUnit",
            "UnitPrice",
        ]
        float_col_list = ["Frontage", "Breadth"]

        def _to_int(x: str):
            """
            数値型に変更する関数
            nanをはじきたい
            """
            try:
                x = int(x)
                return x
            except:
                return x

        for col in int_col_list:
            if col == "Area":
                df[col] = df[col].map(lambda x: x.replace("㎡以上", ""))
            df[col] = df[col].map(_to_int)

        for col in float_col_list:
            df[col] = df[col].astype(float, errors="ignore")

        return df

    def _wareki_to_seireki(self, x: Union[str, float]) -> int:
        if not x or isinstance(x, float) and np.isnan(x):
            return np.nan
        try:
            if x[:2] == "令和":
                year = int(x[2:-1])
                wareki = 2018 + year
                return wareki
            elif x[:2] == "平成":
                year = int(x[2:-1])
                wareki = 1988 + year
                return wareki
            elif x[:2] == "昭和":
                year = int(x[2:-1])
                wareki = 1925 + year
                return wareki
        except ValueError:
            return np.nan

    def _convert_to_quarter(self, input_str: str) -> pd.Period:
        # 年と四半期の情報を抽出
        year = int(input_str.split("年")[0])
        quarter = int(input_str.split("第")[1][0])
        return pd.Period(f"{year}Q{quarter}")


if __name__ == "__main__":
    t = CityCode("京都")
    pref_code = t.pref_code
    city_code = t.city_code("京都市")
    kt = PropTransactions(pref_code, city_code, 20223, 20231)
    df = kt.get_data()
    print(df.head())


j_realty_api/j_realty_en.py
import json
import math
from dataclasses import dataclass
from datetime import datetime
from typing import Union

import numpy as np
import pandas as pd
import pkg_resources
import requests
from JPYForex import jpyforex

# API from 国土交通省
# web: https://www.land.mlit.go.jp/webland/api.html

CITYSEARCH_EN_URL = "https://www.land.mlit.go.jp/webland_english/api/CitySearch"
TRADESEARCH_EN_URL = "https://www.land.mlit.go.jp/webland_english/api/TradeListSearch"
EN_PREF_CODE_JSON_PATH = pkg_resources.resource_filename(
    "j_realty_api.j_realty_en", "data/pref-code-en.json"
)


@dataclass
class CityCode:

    """
    A class that, when given a prefecture name, returns data containing
    the names and codes of the municipalities in that prefecture. When a
    municipality name is specified, it returns the municipality code.

    """

    pref_name: str
    base_path: str = CITYSEARCH_EN_URL

    @property
    def pref_code(self) -> str:
        """
        search pref_code from JSON with pref_name
        Returns:
            str: pref_code
        """
        with open(EN_PREF_CODE_JSON_PATH, "r", encoding="utf-8") as f:
            pref_code_json = json.load(f)
        for p_name, p_code in pref_code_json.items():
            if p_name.startswith(self.pref_name.lower()):
                return p_code
        raise Exception(f"No pref_code found for {self.pref_name}")

    @property
    def city_json(self) -> dict[str]:
        """
        api returns each pref's city_codes json.
        Returns:
            Dict[str]: JSON with {'cityname' : 'citycode'}
        """
        r = requests.get(self.base_path, params={"area": self.pref_code})
        city_json = r.json()
        new_list = list()
        for d in city_json["data"]:
            d["name"] = d["name"].lower()
            new_list.append(d)
        city_json["data"] = new_list
        return city_json

    def city_code(self, city_name: str) -> str:
        """
        Searches for the city code corresponding to the provided
        city name from the city_json.

        Args:
            city_name str: City name for which to find the city code.
        Returns:
            str: city code

        """
        for d in self.city_json.get("data", []):
            if d["name"].startswith(city_name.lower()):
                return d["id"]
        raise Exception(f"No city_code found for {city_name}.")


@dataclass
class PropTransactions:

    """
    A class that retrieves property transaction data based on the specified parameters
    such as prefecture code, city code, and date range from a given API endpoint.
    """

    pref_code: str
    city_code: str
    from_dt: int = 20222
    to_dt: int = 20223
    base_url: str = TRADESEARCH_EN_URL

    def get_data(self) -> pd.DataFrame:
        """
        Sends a GET request to the API with the specified parameters and retrieves
        property transaction data.

        The method constructs a DataFrame from the received JSON data and returns it.

        Returns:
            pd.DataFrame: A DataFrame containing the property transaction data.

        """
        params = {
            "from": self.from_dt,
            "to": self.to_dt,
            "area": self.pref_code,
            "city": self.city_code,
        }
        r = requests.get(self.base_url, params=params)
        if r.status_code == 200:
            try:
                df = pd.DataFrame(r.json()["data"])
                df["Period"] = df["Period"].map(self.custom_to_datetime)
                df["TradePrice"] = df["TradePrice"].astype("Int64")
                df["UnitPrice"] = pd.to_numeric(df["UnitPrice"], errors="ignore")
                return df
            except KeyError:
                print(
                    f"""KeyError: The received JSON does not contain the expected "data" key.
                      pref_code: {self.pref_code}, city_code: {self.city_code}, 
                      from_dt: {self.from_dt}, to_dt: {self.to_dt}
                      """
                )
        else:
            raise Exception(f"Status_code: {r.status_code}")

    def to_forex(self, df: pd.DataFrame, currency: str) -> pd.DataFrame:
        """
        Params:
            df: pd.DataFrame
                DataFrame from get_data()
            currency: str
                select from 3types: "USD", "EUR", "CNY"

        Returns:
            pd.DataFrame
        """
        dates = df["Period"].unique()
        min_date = dates.min().to_timestamp().date()
        max_date = dates.max().to_timestamp(how="End").date()
        min_date_str = self.dt_to_string(min_date)
        max_date_str = self.dt_to_string(max_date)
        cur = jpyforex.JPYForex(
            currency=currency, freq="Q", start_date=min_date_str, end_date=max_date_str
        )
        data = cur.get_data()
        data.index = data.index.to_period("Q")
        forex_data = dict()
        for i in data.index:
            forex_data[i] = data.loc[i].values[0]
        df["forex"] = df["Period"].map(forex_data)
        df["ForexTradePrice"] = df.apply(lambda x: x["TradePrice"] / x["forex"], axis=1)
        df["ForexUnitPrice"] = df.apply(lambda x: x["UnitPrice"] / x["forex"], axis=1)
        return df

    def custom_to_datetime(self, p_str):
        split_str = p_str.split(" ")
        year = int(split_str[2])
        q = split_str[0]
        if q == "1st":
            qt = self.dt_to_qt(year, 2)
            return qt
        elif q == "2nd":
            qt = self.dt_to_qt(year, 5)
            return qt
        elif q == "3rd":
            qt = self.dt_to_qt(year, 8)
            return qt
        elif q == "4th":
            qt = self.dt_to_qt(year, 11)
            return qt
        else:
            raise ValueError(f"invalid quater: {p_str}")

    def dt_to_qt(self, year, month):
        dt = datetime(year, month, 1)
        ts = pd.Timestamp(dt)
        qt = ts.to_period("Q")
        return qt

    def dt_to_string(self, dt: datetime) -> str:
        year = dt.year
        month = dt.month
        date = dt.day
        str_dt = f"{year}{month:02}{date:02}"
        return str_dt


if __name__ == "__main__":
    k = CityCode("Kyoto")
    pref_code = k.pref_code
    city_code = k.city_code("Kyoto")
    print(pref_code)
    print(city_code)
